workflow/ttlcontroller/ttlcontroller.go

```diff
@@ -137,121 +103,69 @@ func (c *Controller) enqueueWF(obj interface{}) {
 		log.Warnf("Failed to unmarshal workflow %v object: %v", obj, err)
 		return
 	}
-	now := c.clock.Now()
-	remaining, expiration := timeLeft(wf, &now)
-	if remaining == nil || *remaining > c.resyncPeriod {
-		return
-	}
-	log.Infof("Found Workflow %s/%s set expire at %v (%s from now)", wf.Namespace, wf.Name, expiration, remaining)
-	var addAfter time.Duration
-	if *remaining > 0 {
-		addAfter = *remaining
-	}
-	var key string
-	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
-		runtimeutil.HandleError(err)
+	remaining, ok := c.expiresIn(wf)
+	if !ok {
 		return
 	}
 	// if we try and delete in the next second, it is almost certain that the informer is out of sync. Because we
 	// double-check that sees if the workflow in the informer is already deleted and we'll make 2 API requests when
 	// one is enough.
-	if addAfter < enoughTimeForInformerSync {
-		addAfter = enoughTimeForInformerSync
-	}
-	log.Infof("Queueing workflow %s/%s for delete in %v", wf.Namespace, wf.Name, addAfter)
+	// Additionally, this allows enough time to make sure the double checking that the workflow is actually expired
+	// truly works.
+	addAfter := remaining + time.Second
+	key, _ := cache.MetaNamespaceKeyFunc(obj)
+	log.Infof("Queueing %v workflow %s for delete in %v", wf.Status.Phase, key, addAfter.Truncate(time.Second))
 	c.workqueue.AddAfter(key, addAfter)
 }
 
 func (c *Controller) deleteWorkflow(key string) error {
-	obj, exists, err := c.wfInformer.GetIndexer().GetByKey(key)
```

Comment:
this is almost a re-write to simplify - if you add a workflow to the TTL queue - we assume it must be deleted